### DISCUSSION TITLE
Plans for v2.0.0

### DISCUSSION THREAD

**Discussion by empicano** [empicano/aiomqtt#226] [General]
This discussion serves to collect ideas/plans for breaking changes for aiomqtt's version 2.0.0 💥 Ideas are welcome!

Some ideas:

- Switch to a client-wide queue to simplify the API as explained [in this comment](https://github.com/sbtinstruments/aiomqtt/pull/220#issuecomment-1592041102) (this includes creating documentation about how to implement the multiple queues back on top)
- Remove the deprecated `filtered_messages` and `unfiltered_messages` context managers
- Remove `connect`/`disconnect` methods (`__aenter__` and `__aexit__` can be used as a fallback, but it's more obvious that one shouldn't)
- [Rename `id` to `identifier`](https://github.com/sbtinstruments/aiomqtt/blob/f8fc6669f446ebf846522318bc3c96f6805efe94/aiomqtt/client.py#L381) (or similar)

**Comment by frederikaalund** [comment]
Great with this discussion! 😄 Here are some of my thoughts (not plans, just ideas):

 * Switch to anyio internally. Simplifies the implementation a lot. Less edge cases. No more raw tasks and futures.
 * Separate the aiomqtt API into a "user friendly" layer on top of a "strict" layer. The latter uses the [Typestate Pattern](http://cliffle.com/blog/rust-typestate/) known from rust ([see this blog post for a Python example](https://kobzol.github.io/rust/python/2023/05/20/writing-python-like-its-rust.html#encoding-invariants-using-types)). This separation makes it much easier to maintain (and test!) aiomqtt.

I also like to note that at this point, aiomqtt is mostly an exercise in idiomatic API design. It lacks the raw, concurrency-based performance that users expect from an "aio"-prefixed library. The problem is that we use paho-mqtt underneath that, in turn, use the blocking socket API. See #205 for reference. To get around this limitation, we can introduce alternative "back-ends" in addition to paho-mqtt. On the top of my head: [gmqtt](https://github.com/wialon/gmqtt) and [amqtt](https://github.com/Yakifo/amqtt) (fork of hbmqtt).

Another alternative back-end: Wrapper around the native [rumqtt](https://github.com/bytebeamio/rumqtt) client. It uses async I/O underneath.

**Reply by empicano** [reply]
Very interesting! The blog post you linked explaining how to adapt a client class fits almost too well 😄 I like the idea of the Typestate Pattern; At first glance, I wouldn't know how to represent a client's connection loss with it. Do you have any ideas?

> aiomqtt is mostly an exercise in idiomatic API design

That's very critical 😄 But you're right, "real" concurrency is definitely missing 👍

I'd prefer to go all in on a single backend instead of supporting multiple. Abstractions like this often end up leaky. For example, SQLAlchemy implements this unified "core" language but then still needs dialects for the different databases (which still doesn't allow everything, but that's beside the point). In addition, I think there's little benefit that would justify the increased maintenance burden. Happy to be proven wrong though! :)

rumqtt sounds the most interesting to me as a new backend, also because it seems to be the one that's most actively maintained. [It lacks some MQTT features](https://github.com/bytebeamio/rumqtt#-rumqttd), but gmqtt and amqtt do as well. Maybe we can switch to rumqtt and use paho-mqtt as a fallback until it catches up with the features. @JonathanPlasse gave [a rust backend a try](https://github.com/JonathanPlasse/mqttbytes_python) a while ago, I wonder what he thinks about this?

**Reply by JonathanPlasse** [reply]
I did a partial wrapper around [mqttbytes](https://github.com/bytebeamio/mqttbytes) used by rumqtt internally.
It only implements the assembling and disassembling of MQTT messages.
I could try to make a wrapper of rumqtt.

**Reply by frederikaalund** [reply]
> I did a partial wrapper around [mqttbytes](https://github.com/bytebeamio/mqttbytes) used by rumqtt internally. It only implements the assembling and disassembling of MQTT messages. I could try to make a wrapper of rumqtt.

Sounds interesting. :) Like a [sans I/O](https://sans-io.readthedocs.io/) approach?

**Reply by JonathanPlasse** [reply]
Exactly

**Comment by ihavenonickname** [comment]
I searched in the documentation for a more concise way to subscribe to a topic, but found nothing. 

See:

```python
async def main():
    async with aiomqtt.Client("localhost") as client:
        async with client.messages() as messages:
            await client.subscribe("temperature/#")
            async for message in messages:
                ...
```

That's a lot of code.

It would be nice if we had a cleaner, shorter way to do such a common operation, like so:

```python
async def main():
    async for message in aiomqtt.subscribe('localhost', 1883, 'temperature/#'):
        ...

```

That's similar to the nice [`mosquitto_sub` CLI](https://mosquitto.org/man/mosquitto_sub-1.html):

```bash
mosquitto_sub -h localhost -p 1883 -t temperature/#
```

**Reply by m-u-xyz** [reply]
Well, opening a new connection for each subscription seems somewhat wasteful, but the salient point is that the user shouldn't be forced to write their own message dispatcher.

Thus we should at least be able to do this:
```python3
async def main():
    async with aiomqtt.Client("localhost") as client:
        async with client.subscribe("temperature/#") as messages:
            async for message in messages:
                ...
```
I implemented this in my own anyio-ified `hbmqtt` fork (`moat-mqtt`) but hbmqtt is a dead end these days.

Too bad that MQTT doesn't have a subscription IDs and thus forces us to walk through the topics on each message.
